[TTS] bug fix - sample rate was being ignored in vocoder dataset

paarthneekhara opened the pull request:
Signed-off-by: Paarth Neekhara <paarth.n@gmail.com>

# What does this PR do ?

Fixes a bug where sample rate was being ignored in VocoderDataset when loading just the audio and audio lengths and not the precomputed mel.

**Collection**: tts

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
dataset = VocoderDataset(manifest_filepath=manifest_path, sample_rate=16000)
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.


## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

redoctopus reviewed (approved):
LGTM

XuesongYang commented:
@paarthneekhara, thanks for your PR! please address this error: 
```
  File "/opt/conda/lib/python3.8/site-packages/nemo/collections/tts/losses/hifigan_losses.py", line 69, in forward

    loss += torch.mean(torch.abs(rl - gl))

RuntimeError: The size of tensor a (1882) must match the size of tensor b (1878) at non-singleton dimension 2
```

paarthneekhara commented:
@XuesongYang @redoctopus I think I fixed the issue. Adjusted the number of segments to get at the original sampling rate so that we get n_segments at our target sampling rate. Let me know if this looks ok.

redoctopus commented on nemo/collections/asr/parts/preprocessing/segment.py:
```diff
@@ -186,19 +187,27 @@ def segment_from_file(cls, audio_file, target_sr=None, n_segments=0, trim=False,
-                if 0 < n_segments < len(f):
-                    max_audio_start = len(f) - n_segments
+                if target_sr is not None:
+                    n_segments_at_original_sr = math.ceil(n_segments * sample_rate / target_sr)
```
As mentioned in-person--please double-check the test case for both upsampling and downsampling

XuesongYang replied:
Out of curiosity, why do we have to handle the case that `sample_rate` mismatches `target_sr` in the func `sefment_from_file`? In the return `cls()`, resampling would be applied as shown in https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/parts/preprocessing/segment.py#L98.

redoctopus replied:
IIRC Paarth was running into an issue where the mismatch caused an unexpected number of segments to be returned since `n_segments` calculated at the original sampling rate would not be the same as for the target sample rate. So this needed to be corrected for before passing in the segments for resampling.

XuesongYang replied:
yes. I remembered that (see my above comment with detailed errors). But one thing that confused me: 
* why `n_segments (num of samples)` was independent to `sample_rate`, and why it was not automatically calculated based on `seconds * sample_rate`? 
* If you look at https://github.com/NVIDIA/NeMo/blob/main/examples/tts/conf/hifigan/hifigan_44100.yaml#L20, it applies for some fixed number as `train_n_segments: 16384`. Do we assume this number applies to any sample rate?

redoctopus replied:
Honestly, I'm not completely sure. But if I had to guess, maybe something to do with GPU memory during training? Similar to how we have a max audio length cutoff, perhaps...

paarthneekhara replied:
Yes, I think for HiFiGAN the train_n_segments should be 16384 for any sampling rate (Probably because it upsamples the spectrogram by some power of 2 at each layer). Also, I think the expected behavior of the function should be to get n_segments at the specified sampling rate.

XuesongYang replied:
Thank you guys for the answers. It helps me understand well. Approved.

redoctopus commented on nemo/collections/asr/parts/preprocessing/segment.py:
```diff
@@ -186,19 +187,27 @@ def segment_from_file(cls, audio_file, target_sr=None, n_segments=0, trim=False,
         samples = samples.transpose()
-        return cls(samples, sample_rate, target_sr=target_sr, trim=trim, orig_sr=orig_sr)
+        features = cls(samples, sample_rate, target_sr=target_sr, trim=trim, orig_sr=orig_sr)
+        features._samples = features._samples[:n_segments]
```
Add case for `n_segments` = 0

redoctopus reviewed (approved):
LGTM

XuesongYang reviewed (changes requested):
Please also fix the conflicts.

redoctopus reviewed (approved):
Fixed conflict, approving again

XuesongYang commented:
It would be great if we have a unit test capturing this bug fix.

XuesongYang reviewed (approved):
LGTM. Thanks for the fix.
